modules/e2eping/e2eping.js

```diff
@@ -0,0 +1,333 @@
+/* global __filename */
+import { getLogger } from 'jitsi-meet-logger';
+import * as JitsiConferenceEvents from '../../JitsiConferenceEvents';
+import { createE2eRttEvent } from '../../service/statistics/AnalyticsEvents';
+import * as E2ePingEvents
+    from '../../service/e2eping/E2ePingEvents';
+import Statistics from '../statistics/statistics';
+
+const logger = getLogger(__filename);
+
+/**
+ * The 'type' of a message which designates an e2e ping request.
+ * @type {string}
+ */
+const E2E_PING_REQUEST = 'e2e-ping-request';
+
+/**
+ * The 'type' of a message which designates an e2e ping response.
+ * @type {string}
+ */
+const E2E_PING_RESPONSE = 'e2e-ping-response';
+
+/**
+ * Saves e2e ping related state for a single JitsiParticipant.
+ */
+class ParticipantState {
+    /**
+     * Creates a ParticipantState
+     * @param {JitsiParticipant} participant
+     * @param {E2ePing} e2eping
+     */
+    constructor(participant, e2eping) {
+        // The JitsiParticipant
+        this.participant = participant;
+
+        // The E2ePing
+        this.e2eping = e2eping;
+
+        // Caches the ID
+        this.id = participant.getId();
+
+        // Recently sent requests
+        this.requests = {};
+
+        // The ID of the last sent request. We just increment it for each new
+        // request. Start at 1 so we can consider only thruthy values valid.
+        this.lastRequestId = 1;
+
+        // If the data channel was already open (this is likely a participant
+        // joining an existing conference) send a request immediately.
+        if (e2eping.dataChannelOpened) {
+            this.sendRequest();
+        }
+
+        this.pingInterval = window.setInterval(
```

Comment:
i was expecting that e2eping would actually be responsible for a lot of the stuff that's being done here.  maybe it's just due to naming, but i was expecting `ParticipantState` to basically just hold data, and `E2eping` to be doing all the management of pings (when they should be sent, the current sequence number, etc.) and would just fire some round trip event or something whenever a new one was calculated.  is there an E2eping instance for each remote participant?  (if so, i wonder if there is anything to the idea of a single broadcast ping that all remote participants would respond to rather than multiple individuals ones?  but maybe that would be harder to manage, i dunno)

Reply:
There is a single E2ePing instance. And one (what is now) ParticipantWrapper per remote partcipant.

Reply:
I've added a TODO for using a broadcast, it's definitely a better solution